content/posts/2019-05-job-search-2-parsing.md

```diff
@@ -0,0 +1,276 @@
+
+    private static Parser<char, Func<JqlNode, JqlNode, JqlNode>> Binary(Parser<char, JqlNodeType> op) =>
+            op.Select<Func<JqlNode, JqlNode, JqlNode>>(type => (l, r) => new BinaryNode(l, r, type));
+
+    private static readonly Parser<char, JqlNode> _expressionParser = 
+         ExpressionParser.Build<char, JqlNode>(
+            p => OneOf(
+                    _tag,
+                    Parenthesised(p)
+            ).AtLeastOnce().Select<JqlNode>(JqlBuilder.Group),
+            new[]
+            {
+                Operator.InfixL(_and),
+                Operator.InfixL(_or)
+            }
+        ).AtLeastOnce().Select<JqlNode>(x => new QueryNode(x));
+
+    public static QueryNode Parse(string input) => (QueryNode)_expressionParser.ParseOrThrow(input);
+}
+```
+
+Here we use a Pidgin helper called `ExpressionParser.Build` that handles a lot of the pain associated with parsing more complex expressions;
+
+ - Its first parameter takes a `Func<Parser, Parser>` that returns a parser that can be used for parsing individual terms in the expression. Here we say that we either accept a tag *or* something that matches any term defined here wrapped in parenthesis. This is important because it allows us to trivially handle recursion in the grammar.
+ - Its second parameter takes an array of operators that can be applied to individual terms in the expression. We support `and` and `or` and we apply them using [infix](http://www.cs.man.ac.uk/~pjj/cs212/fix.html) notation from the left-hand side of the expression. That means we that we treat `[a] and [b] or [c]` as `([a] and [b]) or [c]` rather than `[a] and ([b] or [c])`.
+ - Our `and` and `or` parsers are defined as functions that take previously parsed `JqlNode` objects and combine them into `BinaryNode` objects with the correct operator.
+
+ That's it - we can throw input at this thing and it'll be parsed to its equivalent AST representation.
+
+But that's not the end of the tale, we have a grammar that works well for *expected* input, but production web apps end up encountering all kinds of random junk... How do we make our parser resilient to this kind of input?
+
+## Handling Bad Input
+
+We've seen all kinds of nonsense make its way into job search. Sometimes it's typos, other times it's malicious input - how should we handle this kind of input? We *could* just return a HTTP `400 Bad Request` with details of where in the input we failed but we can generally pull something useful enough out of the input to run a query against the backend, so why not do so?
+
+Our most common fallback is to treat anything we don't really understand as just a text query. It's a reasonable fallback - we return results and users generally adjust their query, possibly correcting syntactic mistakes, if the results don't seem to be that useful. The way we treat text queries in Elastic means that we'll generally pick up anything useful in the user's input.
```

Comment:
Which way is that? If we’re not elaborating, Id remove that sentence for the sake of clarity as well